lib/redis_memo/memoize_query/cached_select/bind_params.rb

```diff
@@ -55,73 +115,68 @@ def product(other)
-        # Distribute the current attrs into the other
-        other_attrs_set_size = other.params[model].size
-        other_attrs_set = other.params[model]
-        merged_attrs_set = Array.new(other_attrs_set_size * attrs_set.size)
-
-        attrs_set.each_with_index do |attrs, i|
-          other_attrs_set.each_with_index do |other_attrs, j|
-            k = i * other_attrs_set_size + j
-            merged_attrs = merged_attrs_set[k] = other_attrs.dup
+        # distribute the current attrs into the other
+        merged_attrs_set = Set.new
+        attrs_set.each do |attrs|
+          other_attrs_set.each do |other_attrs|
+            merged_attrs = other_attrs.dup
+            should_add = true
             attrs.each do |name, val|
-              # Conflict detected. For example:
+              # conflict detected. for example:
               #
               #   (a = 1 or b = 1) and (a = 2 or b = 2)
               #
-              #  Keep:     a = 1 and b = 2, a = 2 and b = 1
-              #  Discard:  a = 1 and a = 2, b = 1 and b = 2
+              #  keep:     a = 1 and b = 2, a = 2 and b = 1
+              #  discard:  a = 1 and a = 2, b = 1 and b = 2
               if merged_attrs.include?(name) && merged_attrs[name] != val
-                merged_attrs_set[k] = nil
+                should_add = false
                 break
               end
 
               merged_attrs[name] = val
             end
+            merged_attrs_set << merged_attrs if should_add
           end
         end
 
-        merged_attrs_set.compact!
-        other.params[model] = merged_attrs_set
+        merged_attrs_set
       end
+    end
 
-      other
+    def plan_product
+      plan.dependency_size = left.plan.dependency_size * right.plan.dependency_size
+      plan.model_attrs = product_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
     end
 
-    def uniq!
-      params.each do |_, attrs_set|
-        attrs_set.uniq!
-      end
+    def product!
+      @params = product_attrs_set(left.params, right.params)
     end
 
-    def memoizable?
-      return false if params.empty?
+    class Plan
+      attr_accessor :dependency_size
+      attr_accessor :model_attrs
 
-      params.each do |model, attrs_set|
-        return false if attrs_set.empty?
+      def initialize(bind_params)
+        @dependency_size = 0
+        @model_attrs = Hash.new do |models, model|
+          models[model] = Set.new
+        end
 
-        attrs_set.each do |attrs|
-          return false unless RedisMemo::MemoizeQuery
-            .memoized_columns(model)
-            .include?(attrs.keys.sort)
+        # An aggregated bind_params node can only obtain params by combining
+        # its children nodes
+        return if !bind_params.__send__(:opt).nil?
+
+        bind_params.params.each do |model, attrs_set|
+          @dependency_size += attrs_set.size
+          attrs_set.each do |attrs|
+            @model_attrs[model] << attrs.keys.map { |k| [k, nil] }.to_h
```

Comment:
can we explain why we passed in `nil` here? also related to https://github.com/chanzuckerberg/redis-memo/pull/94/files#r650273177, I think using a set seems to make more sense here, so if the advantage of using a Hash here is to avoid another code path, I think we should add some explanations here